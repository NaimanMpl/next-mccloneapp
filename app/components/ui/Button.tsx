import React from 'react';

interface ButtonProps {
  label: string,
  backgroundColor: string,
  color: string,
  fullWidth?: boolean,
  loading?: boolean,
  submit?: boolean
}

const Button = ({ backgroundColor, color, loading = false, label, fullWidth = true, submit = true }: ButtonProps) => {
  const textColor = `text-${color}`;
  return (
    <button
      type={submit ? 'submit' : undefined}
      className={`mt-4 ${fullWidth ? 'w-full' : 'px-10'} ${backgroundColor} rounded-md py-2 font-semibold ${textColor}`}
      disabled={loading}
    >
      {label}
    </button>
  )
}

export default Button;