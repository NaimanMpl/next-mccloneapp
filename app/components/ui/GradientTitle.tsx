import { cn } from '@/lib/utils'
import { ReactNode } from 'react'

interface GradientTitleProps {
  children: ReactNode,
  className?: string
}

const GradientTitle = ({ children, className }: GradientTitleProps) => {
  return (
    <h2 
      className={cn(`pointer-events-none z-10 whitespace-pre-wrap bg-gradient-to-r from-[#000000] from-65% to-neutral-200/60 bg-clip-text text-center text-7xl font-semibold leading-none text-transparent dark:from-white dark:to-slate-900/10 mobile:text-5xl mobile:from-50%`, className)}
    >
      {children}
    </h2>
  )
}

export default GradientTitle