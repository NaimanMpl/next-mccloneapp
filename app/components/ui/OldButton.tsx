import React from 'react';

interface OldButtonProps {
  label: string,
  backgroundColor: string,
  color: string,
  fullWidth?: boolean,
  loading?: boolean,
  submit?: boolean
}

const OldButton = ({ backgroundColor, color, loading = false, label, fullWidth = true, submit = true }: OldButtonProps) => {
  const textColor = `text-${color}`;
  return (
    <button
      type={submit ? 'submit' : undefined}
      className={`${fullWidth ? 'w-full' : 'px-10'} ${backgroundColor} rounded-md py-2 font-semibold ${textColor}`}
      disabled={loading}
    >
      {label}
    </button>
  )
}

export default OldButton;