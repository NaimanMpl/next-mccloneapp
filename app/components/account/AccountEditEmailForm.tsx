import { useEditEmailForm } from '@/app/hooks/useEditEmailForm';
import { Button } from '@/components/ui/button';
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from '@/components/ui/card';
import { Form, FormControl, FormField, FormItem, FormMessage } from '@/components/ui/form';
import { Input } from '@/components/ui/input';
import { Loader2 } from 'lucide-react';

interface AccountEditEmailFormProps {
  email: string
}

const AccountEditEmailForm = ({ email }: AccountEditEmailFormProps) => {

  const { form, onSubmit, loading } = useEditEmailForm({
    defaultValue: email
  });

  return (
    <Card>
      <CardHeader>
        <CardTitle className='text-xl font-semibold'>Adresse mail</CardTitle>
        <CardDescription>Veuillez saisir une adresse email avec laquelle vous souhaitez pouvoir vous connecter.</CardDescription>
      </CardHeader>
      <Form {...form}>
        <form onSubmit={form.handleSubmit(onSubmit)}>
          <CardContent>
            <FormField 
              control={form.control}
              name='email'
              render={({ field }) => (
                <FormItem>
                  <FormControl>
                    <Input className='mt-3 w-80' type='text' defaultValue={email} {...field} />
                  </FormControl>
                  <FormMessage />
                </FormItem>
              )}
            />
          </CardContent>
          <CardFooter className='border-t border-border px-6 py-2'>
            <div className='flex justify-between items-center w-full'>
              <CardDescription>Les adresses emails doivent êtres vérifiés afin de pouvoir se connecter avec.</CardDescription>
              <Button type='submit' disabled={loading}>
                {loading && <Loader2 className='mr-2 h-4 w-4 animate-spin' />}
                Sauvegarder
              </Button>
            </div>
          </CardFooter>
        </form>
      </Form>
    </Card>
  )
}

export default AccountEditEmailForm