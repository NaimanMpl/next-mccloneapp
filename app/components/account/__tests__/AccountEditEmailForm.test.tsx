import mockAxios, { SESSION_MOCK } from "@/__mocks__/axios";
import { Toaster } from "@/components/ui/toaster";
import { render, screen, waitFor } from "@testing-library/react";
import user from '@testing-library/user-event';
import { useSession } from "next-auth/react";
import AccountEditEmailForm from "../AccountEditEmailForm";

jest.mock('axios');
jest.mock('next-auth/react');

describe('<AccountEditEmailForm>', () => {

  beforeEach(() => {
    (useSession as jest.Mock).mockReturnValue({ data: SESSION_MOCK, update: jest.fn() });
  })

  afterEach(() => {
    mockAxios.reset();
    user.setup();
  })

  it('should render successfully', () => {

    const email = SESSION_MOCK.user.email;

    const form = render(
      <AccountEditEmailForm email={email} />
    );

    const [ emailInput ] = screen.getAllByRole('textbox');

    expect(emailInput).toHaveValue(email);
    expect(form).toMatchSnapshot();
  });

  it('should change input value on successfull email rename api call', async () => {
    
    const newEmail = "john.doe2@domain.com";
    const email = SESSION_MOCK.user.email;
    mockAxios.patch.mockResolvedValue({ username: newEmail });

    render(
      <AccountEditEmailForm email={email} />
    );

    const [ emailInput ] = screen.getAllByRole('textbox');

    await user.clear(emailInput);
    await user.type(emailInput, newEmail);

    const submitButton = screen.getByText('Sauvegarder');
    await user.click(submitButton);

    await waitFor(() => {
      expect(mockAxios.patch).toHaveBeenCalledWith(`/api/auth/me?email=${newEmail}`);
      expect(emailInput).toHaveValue(newEmail);
    });
    
  });

  it('should throw a warning toast if the user keeps his email unchanged', async () => {

    const email = SESSION_MOCK.user.email;

    render(
      <>
        <AccountEditEmailForm email={email} />
        <Toaster />
      </>
    );

    const [ emailInput ] = screen.getAllByRole('textbox');

    await user.clear(emailInput);
    await user.type(emailInput, email);

    const submitButton = screen.getByText('Sauvegarder');
    await user.click(submitButton);

    await waitFor(() => {
      expect(emailInput).toHaveValue(email);
      expect(screen.getByText('Pas si vite !')).toBeInTheDocument();
    });
    
  });

  it('should throw an error if the user try to submit an invalid email', async () => {

    const email = SESSION_MOCK.user.email;
    const newEmail = "not_a_valid_email";

    render(
      <>
        <AccountEditEmailForm email={email} />
      </>
    );

    const [ emailInput ] = screen.getAllByRole('textbox');

    await user.clear(emailInput);
    await user.type(emailInput, newEmail);

    const submitButton = screen.getByText('Sauvegarder');
    await user.click(submitButton);

    await waitFor(() => {
      expect(emailInput).toHaveValue(newEmail);
      expect(screen.getByText('Veuillez renseigner une adresse mail valide')).toBeInTheDocument();
    });
    
  });

});