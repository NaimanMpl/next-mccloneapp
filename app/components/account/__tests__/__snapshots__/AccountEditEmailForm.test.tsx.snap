// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountEditEmailForm> should render successfully 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="rounded-lg border border-border bg-card text-card-foreground shadow-sm"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <h3
            class="tracking-tight text-xl font-semibold"
          >
            Adresse mail
          </h3>
          <p
            class="text-sm text-muted-foreground"
          >
            Veuillez saisir une adresse email avec laquelle vous souhaitez pouvoir vous connecter.
          </p>
        </div>
        <form>
          <div
            class="p-6 pt-0"
          >
            <div
              class="space-y-2"
            >
              <input
                aria-describedby=":r0:-form-item-description"
                aria-invalid="false"
                class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
                id=":r0:-form-item"
                name="email"
                type="text"
                value="johndoe@domain.com"
              />
            </div>
          </div>
          <div
            class="flex items-center p-6 border-t border-border px-6 py-2"
          >
            <div
              class="flex justify-between items-center w-full"
            >
              <p
                class="text-sm text-muted-foreground"
              >
                Les adresses emails doivent êtres vérifiés afin de pouvoir se connecter avec.
              </p>
              <button
                class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-semibold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
                type="submit"
              >
                Sauvegarder
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="rounded-lg border border-border bg-card text-card-foreground shadow-sm"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <h3
          class="tracking-tight text-xl font-semibold"
        >
          Adresse mail
        </h3>
        <p
          class="text-sm text-muted-foreground"
        >
          Veuillez saisir une adresse email avec laquelle vous souhaitez pouvoir vous connecter.
        </p>
      </div>
      <form>
        <div
          class="p-6 pt-0"
        >
          <div
            class="space-y-2"
          >
            <input
              aria-describedby=":r0:-form-item-description"
              aria-invalid="false"
              class="flex h-10 rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 mt-3 w-80"
              id=":r0:-form-item"
              name="email"
              type="text"
              value="johndoe@domain.com"
            />
          </div>
        </div>
        <div
          class="flex items-center p-6 border-t border-border px-6 py-2"
        >
          <div
            class="flex justify-between items-center w-full"
          >
            <p
              class="text-sm text-muted-foreground"
            >
              Les adresses emails doivent êtres vérifiés afin de pouvoir se connecter avec.
            </p>
            <button
              class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-semibold ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2"
              type="submit"
            >
              Sauvegarder
            </button>
          </div>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
